### PR TITLE
cmd, common, node, rpc: move HTTP RPC into node, drop singleton aspect

### PR DESCRIPTION
**Pull request by karalabe**
This PR is the continuation of the RPC API integration into the `node`.

 - [x] Break the hard coded singleton nature of the HTTP RPC server
 - [x] Move the HTTP RPC server into the `node`
 - [x] Implement HTTP RPC starting based on config values, not only through the API
 - [x] Break the hard coded singleton nature of the WebSocket RPC server
 - [x] Move the WebSocket RPC server into the `node`
 - [x] Implement WebSocket RPC starting based on config values, not only through the API
 - [x] Remove RPC server and client boilerplate from the codebase
 - [x] Moved the in-process RPC client to `rpc`, exposed via `node`

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Fri Feb 5 11:49:15 UTC 2016_

File: node/config.go [hunk 1]
```diff
@@ -97,6 +98,44 @@ type Config struct {
 	// handshake phase, counted separately for inbound and outbound connections.
 	// Zero defaults to preset values.
 	MaxPendingPeers int
+
+	// HttpHost is the host interface on which to start the HTTP RPC server. If this
+	// field is empty, no HTTP API endpoint will be started.
+	HttpHost string
```
Comment (bas-vk) [inline node/config.go]: Could you change this to HTTPHost (same for the other fields)?
We use this [convention](https://github.com/golang/go/wiki/CodeReviewComments#initialisms) on most places.

Comment (karalabe) [inline node/config.go]: Sure!

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -271,20 +270,20 @@ var (
...
+		Value: common.DefaultWsPort,
 	}
 	WSApiFlag = cli.StringFlag{
 		Name:  "wsapi",
 		Usage: "API's offered over the WS-RPC interface",
 		Value: rpc.DefaultHttpRpcApis,
 	}
-	WSAllowedDomainsFlag = cli.StringFlag{
-		Name:  "wsdomains",
+	WSCORSDomainFlag = cli.StringFlag{
```
Comment (bas-vk) [inline cmd/utils/flags.go]: The websocket standard doesn't support CORS as the HTTP standard does. Therefore the current implementation gives the user the option to specify a list of domains from which websockets are allowed. This check is server side (HTTP CORS is enforced by the browser). Therefore both are not the same and I would recommend leaving the old name in place.

Comment (karalabe) [inline cmd/utils/flags.go]: Ah I see, sorry for the confusion. Will revert it back.

File: cmd/geth/main.go [hunk 3]
```diff
@@ -425,8 +425,10 @@ func console(ctx *cli.Context) {
 	startNode(ctx, node)
 
 	// Attach to the newly started node, and either execute script or become interactive
-	client := utils.NewInProcRPCClient(node)
-
+	client, err := utils.NewRemoteRPCClientFromString("ipc:" + node.IpcEndpoint())
+	if err != nil {
+		utils.Fatalf("Failed to attach to the inproc geth: %v", err)
```
Comment (bas-vk) [inline cmd/geth/main.go]: If a user wants to run the console it requires the IPC interface to be active (can be disabled with `--ipcdisable`). This dependency is a bit strange. I think there is a valid use case for users who would like to run the console without any RPC interface enabled. E.g. platforms which don't have support for unix sockets/named pipes, or the user has no permissions to create one (SELinux).

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -482,6 +481,24 @@ func MakeNAT(ctx *cli.Context) nat.Interface {
...
+func MakeHttpRpcHost(ctx *cli.Context) string {
+	if !ctx.GlobalBool(RPCEnabledFlag.Name) {
+		return ""
+	}
+	return ctx.GlobalString(RPCListenAddrFlag.Name)
+}
+
+// MakeWsRpcHost creates the WebSocket RPC listener interface string from the set
+// command line flags, returning empty if the HTTP endpoint is disabled.
+func MakeWsRpcHost(ctx *cli.Context) string {
```
Comment (bas-vk) [inline cmd/utils/flags.go]: I prefer `MakeWSRPCListener` (same for the function on line 486).

File: cmd/geth/main.go (hunk 3 above)
Comment (karalabe) [inline cmd/geth/main.go]: Seems I've misunderstood in the inproc implementation. For some reason I thought it was HTTP based. However, this could then itself be a very useful addition to the `rpc` package and also exposed via the `node`, to allow interfacing the API via libraries without the need for any external channels. I'll move it to the `rpc` package and work it into the rest of the code.

File: node/api.go [hunk 5]
```diff
@@ -61,83 +58,55 @@ func (api *PrivateAdminAPI) AddPeer(url string) (bool, error) {
...
-	err := rpc.StartWS(address, port, corsDomains, offeredAPIs)
-	return err == nil, err
+	return true, nil
 }
 
 // StopRPC terminates an already running websocket RPC API endpoint.
 func (api *PrivateAdminAPI) StopWS() (bool, error) {
-	err := rpc.StopWS()
-	return err == nil, err
+	api.node.lock.Lock()
```
Comment (bas-vk) [inline node/api.go]: Instead of locking in the API it might be better to lock in the node (e.g. `node.stopWS`) since that is where the race conditions can occur (same for the HTTP interface). I would also move the check if the interface is running to the node.

File: rpc/ipc_windows.go [hunk 6]
```diff
@@ -360,6 +363,9 @@ func Listen(address string) (*PipeListener, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Ugly hack working around some async connectivity issues
+	time.Sleep(100 * time.Millisecond)
```
Comment (bas-vk) [inline rpc/ipc_windows.go]: Could you add a link which described the issue?

**Comment by bas-vk** [comment]
@karalabe, finished the review.

File: cmd/utils/flags.go (hunk 4 above)
Comment (karalabe) [inline cmd/utils/flags.go]: I didn't change this because this is just the host part (i.e. interface/ip address). The port it attached within the node.

File: node/api.go (hunk 5 above)
Comment (karalabe) [inline node/api.go]: The reason I needed to lock in the API is because these start/stop methods assume that the lock is already held. This is because they are using during the huge startup procedure, where the lock's already held. Since these are private methods that cannot be accessed outside the `node` package, I think having explicit access to the lock should be fine. The other solution would be adding a set of wrapper methods into the node that just locks/unlocks and calls these. Similarly the node itself never starts/stops the servers on its own or provides an API for for, so it should be fine to error out in this API struct. I don't see too much value in programatically starting/stopping HTTP/WS endpoints on a node, so that's why I didn't provide public methods to do that.

File: rpc/ipc_windows.go (hunk 6 above)
Comment (karalabe) [inline rpc/ipc_windows.go]: Since adding back the inproc client made tests possible via that channel, I've switched everything over to inproc and obsoleted the IPC usage in tests. Hence the error no longer was relevant and I reverted these back.

We should definitely get to the bottom of this, just figured we can do it in another PR and leave it off from this one.

**Comment by karalabe** [comment]
@bas-vk All issues fixed or replied to.

File: node/api.go (hunk 5 above)
Comment (bas-vk) [inline node/api.go]: Ok, thanks for the explanation and I agree with you about keeping it private and don't becoming a big issue. I don't think that adding wrapper (un)lock methods is really an improvement.

**Comment by bas-vk** [comment]
LGTM :+1:

File: common/defaults.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2016 The go-ethereum Authors
```
Comment (obscuren) [inline common/defaults.go]: This isn't the right place for these type of settings. `common` isn't your average dump place for settings ;-)

Comment (karalabe) [inline common/defaults.go]: Well, I can agree, but the DefaultDataDir and DefaultIpcSocket were always part of common. Soo, along those lines, the rest of the RPC pieces should also go here :D 

Also it's not so easy to remove these, because before my PR, these values here were hard coded at a lot of places into the code separately. Moving them into `cmd/geth` would make them unimportable from other packages. Still, I'm open for any suggestion on how to do this better.

**Comment by obscuren** [comment]
:+1:
